libshvchainpack/js/ws_client.js

```diff
@@ -0,0 +1,166 @@
+const datatoRpcValue = (buff) => {
+	let rd = new ChainPackReader(new UnpackContext(buff));
+	rd.readUIntData()
+	const proto = rd.ctx.getByte();
+	if (proto == Cpon.ProtocolType) {
+		rd = new CponReader(rd.ctx)
+	} else {
+		rd = new ChainPackReader(rd.ctx)
+	}
+	const rpc_val = rd.read();
+	return rpc_val;
+};
+
+const checkOption = (option, name, type) => {
+	if (typeof option !== type) {
+		throw new Error(`Required option ${name} missing`);
+	}
+};
+
+class WsClient {
+	constructor(options) {
+		if (typeof options !== "object") {
+			throw new Error("No options object supplied");
+		}
+
+		this.request_id = 1;
+		this.rpc_handlers = [];
+		this.subscriptions = [];
+		this.request_handler = options.requestHandler || (() => {});
+		this.isUseCpon = options.isUseCpon || false;
+		this.logDebug = options.logDebug || (() => {});
+
+		checkOption(options.user, "user", "string");
+		this.user = options.user;
+
+		checkOption(options.password, "password", "string");
+		this.password = options.password;
+
+		checkOption(options.wsUri, "websocket URI", "string");
+
+		this.websocket = new WebSocket(options.wsUri);
+		this.websocket.binaryType = "arraybuffer";
+
+		this.onConnected = options.onConnected || (() => {});
+		this.onRequest = options.onRequest || (() => {});
+
+		this.websocket.onopen = () => {
+			this.logDebug("CONNECTED");
+			this.callRpcMethod(null, "hello").then(() => {
+				const params = `{"login":{"password":"${this.password}","type":"PLAIN","user":"${this.user}"},"options":{"device":{"mountPoint":"test/websocketosaurus"},"idleWatchDogTimeOut": 60}}`;
+				return this.callRpcMethod(null, "login", params);
+			}).then(() => {
+				this.logDebug("SUCCESS: connected to shv broker");
+				this.onConnected();
+			});
+		};
+
+		this.websocket.onclose = () => {
```

Comment:
if it is not JS coding style, then `onClose` please

Reply:
This is the Websocket api, can't do anything about the name unfortunately